pandas/core/groupby/groupby.py

```diff
@@ -1196,7 +1196,7 @@ def reset_identity(values):
             # when the ax has duplicates
             # so we resort to this
             # GH 14776, 30667
-            if ax.has_duplicates:
+            if ax.has_duplicates and not result.axes[self.axis].equals(ax):
                 indexer, _ = result.index.get_indexer_non_unique(ax.values)
                 indexer = algorithms.unique1d(indexer)
                 result = result.take(indexer, axis=self.axis)
```

Comment:
can add copy=False to reindex on L1204 (as if the ax is the same we don't actually need to copy here)

Reply:
Done